lib/service/service.go

```diff
@@ -5299,7 +5299,26 @@ func initSelfSignedHTTPSCert(cfg *servicecfg.Config) (err error) {
 	}
 	cfg.Log.Warningf("Generating self-signed key and cert to %v %v.", keyPath, certPath)
 
-	creds, err := cert.GenerateSelfSignedCert([]string{cfg.Hostname, "localhost"})
+	var hosts = []string{cfg.Hostname, "localhost"}
+
+	// add web public address hosts to self-signed cert
+	if len(cfg.Proxy.PublicAddrs) > 0 {
+		for _, element := range cfg.Proxy.PublicAddrs {
+			proxyHost, _, err := net.SplitHostPort(element.String())
+			if err != nil {
+				// ignore error since this is a nice to have
+				continue
+			}
+
+			if net.ParseIP(proxyHost) == nil {
+				hosts = append(hosts, proxyHost)
+			} else {
+				cfg.Log.Warningf("Not including ip address %v in self-signed cert.", proxyHost)
+			}
+		}
+	}
```

Comment:
A few suggestions here.

- eliminate the else by continuing early
- fix the log message (this happens if the value is not a valid IP, so we shouldn't refer to it as an IP)

```suggestion
			if ip := net.ParseIP(proxyHost); ip == nil {
				cfg.Log.Warningf("%q is not a valid IP address and will not be included in the proxy's web cert", proxyHost)
				continue
			}
			hosts = append(hosts, proxyHost)
		}
	}
```

Reply:
I've removed this section as I thought this was introducing an error to include IPs in the self-signed cert. The `net.ParseIP` was to check if it was an IP or not.  If an IP skip.  This doesn't actually introduce an error so allowing.  If we don't want to include IPs I can put it back in.

Reply:
Whilst it doesn't emit an error, passing an IP into GenerateSelfSignedCert will result in it being encoded as a DNS SAN on the certificate. I'm not too sure on client behaviour, but I'm not sure this will have the intended effect. We should either skip IP addresses (if configuring proxy_service.public_addrs with an IP doesn't make sense) or introduce a new parameter to `GenerateSelfSignedCert` that will make sure IPs go in as a IP SAN.

Reply:
@strideynet please see the latest update.  This will skip adding IP address as DNS SAN and includes a warning message. I put in the comment that we may want to add that in the future